File: MoneyFlow/frontend/src/types/transaction.ts
export type TransactionType = "income" | "expense";

export interface Transaction {
  id: number;
  date: string;
  type: TransactionType;
  amount: number;
  category?: string;
  description?: string;
  created_at: string;
}

export interface TransactionCreate {
  date: string;
  type: TransactionType;
  amount: number;
  category?: string;
  description?: string;
}

export interface TransactionUpdate {
  date?: string;
  type?: TransactionType;
  amount?: number;
  category?: string | null;
  description?: string | null;
}

export interface DailyStatistics {
  date: string;
  income: number;
  expense: number;
  balance: number;
}

export interface StatisticsResponse {
  period_start: string;
  period_end: string;
  total_income: number;
  total_expense: number;
  balance: number;
  daily_statistics: DailyStatistics[];
}

File: MoneyFlow/backend/app/schemas/transaction.py
from pydantic import BaseModel, Field, field_serializer
from datetime import date, datetime
from typing import Optional


class TransactionBase(BaseModel):
    date: date
    type: str = Field(..., pattern="^(income|expense)$")
    amount: float = Field(..., gt=0)
    category: Optional[str] = None
    description: Optional[str] = None


class TransactionCreate(TransactionBase):
    pass


class TransactionUpdate(BaseModel):
    date: Optional[date] = None
    type: Optional[str] = Field(None, pattern="^(income|expense)$")
    amount: Optional[float] = Field(None, gt=0)
    category: Optional[str] = None
    description: Optional[str] = None


class TransactionResponse(TransactionBase):
    id: int
    created_at: datetime

    @field_serializer('created_at')
    def serialize_created_at(self, value: datetime) -> str:
        return value.isoformat()

    class Config:
        from_attributes = True

